### ISSUE TITLE
common_variables.direct_hits.default_definitions is dangerous (Trac #1967)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1959]
The existence of this variable leads users to write code like the following, which has been found in the wild:

```text
    from icecube.common_variables import direct_hits
    DirectHitsDefs=direct_hits.default_definitions
    DirectHitsDefs.append(direct_hits.I3DirectHitsDefinition("E",-15.,250.))
    tray.AddSegment(direct_hits.I3DirectHitsCalculatorSegment,
        DirectHitsDefinitionSeries=DirectHitsDefs,...)
```

This code works as intended as long as it is invoked exactly once. If called a second time, however, it fails with

> FATAL (CommonVariables): common_variables::direct_hits::CalculateDirectHits: Could not insert direct hits values into I3DirectHitsValuesMap using key "E"!

The reason is that `direct_hits.default_definitions` is being modified, rather than copied, so it ends up with two copies of the 'E' entry, and this affects not only the second `I3DirectHitsCalculatorSegment`, but the first as well. This leads to the extremely confusing situation that the second instance appears to the user to a-temporally interfere with the first. Indeed, this will happen even if the two instances are in different trays.

I think that there are two problems here: First, `I3DirectHitsCalculator` does not sanitize its input, so the error is caught later than it could be, and with a less clear message to the user. That can be addressed with this patch:

```text
Index: python/direct_hits/I3DirectHitsCalculator.py
===================================================================
--- python/direct_hits/I3DirectHitsCalculator.py	(revision 153929)
+++ python/direct_hits/I3DirectHitsCalculator.py	(working copy)
@@ -190,6 +190,16 @@
             'not specified!'%
             (self.name)
         )
+        # check for redundancy in definitions
+        definitions={}
+        for definition in self.dh_definitions:
+            if definition.name in definitions:
+                from icecube.icetray import logging
+                logging.log_fatal(
+                    "Conflicting/duplicate DirectHitsDefinitions:\n " \
+                    +str(definitions[definition.name])+"\n " \
+                    +str(definition))
+            definitions[definition.name]=definition
```

Second, and more seriously, `direct_hits.default_definitions` is global state which users tend to modify accidentally. The best solution I can see is to simply eliminate it, and have users instead call the underlying `get_default_definitions`, which has the correct behavior of returning a new object on each call. This is, unfortunately, a breaking change, so I do not want to make it without feedback. 

Migrated from https://code.icecube.wisc.edu/ticket/1967
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:14:55", 
    "description": "The existence of this variable leads users to write code like the following, which has been found in the wild:\n\n{{{\n    from icecube.common_variables import direct_hits\n    DirectHitsDefs=direct_hits.default_definitions\n    DirectHitsDefs.append(direct_hits.I3DirectHitsDefinition(\"E\",-15.,250.))\n    tray.AddSegment(direct_hits.I3DirectHitsCalculatorSegment,\n        DirectHitsDefinitionSeries=DirectHitsDefs,...)\n}}}\n\nThis code works as intended as long as it is invoked exactly once. If called a second time, however, it fails with\n\n> FATAL (CommonVariables): common_variables::direct_hits::CalculateDirectHits: Could not insert direct hits values into I3DirectHitsValuesMap using key \"E\"!\n\nThe reason is that `direct_hits.default_definitions` is being modified, rather than copied, so it ends up with two copies of the 'E' entry, and this affects not only the second `I3DirectHitsCalculatorSegment`, but the first as well. This leads to the extremely confusing situation that the second instance appears to the user to a-temporally interfere with the first. Indeed, this will happen even if the two instances are in different trays.\n\nI think that there are two problems here: First, `I3DirectHitsCalculator` does not sanitize its input, so the error is caught later than it could be, and with a less clear message to the user. That can be addressed with this patch:\n\n{{{\nIndex: python/direct_hits/I3DirectHitsCalculator.py\n===================================================================\n--- python/direct_hits/I3DirectHitsCalculator.py\t(revision 153929)\n+++ python/direct_hits/I3DirectHitsCalculator.py\t(working copy)\n@@ -190,6 +190,16 @@\n             'not specified!'%\n             (self.name)\n         )\n+        # check for redundancy in definitions\n+        definitions={}\n+        for definition in self.dh_definitions:\n+            if definition.name in definitions:\n+                from icecube.icetray import logging\n+                logging.log_fatal(\n+                    \"Conflicting/duplicate DirectHitsDefinitions:\\n \" \\\n+                    +str(definitions[definition.name])+\"\\n \" \\\n+                    +str(definition))\n+            definitions[definition.name]=definition\n}}}\n\nSecond, and more seriously, `direct_hits.default_definitions` is global state which users tend to modify accidentally. The best solution I can see is to simply eliminate it, and have users instead call the underlying `get_default_definitions`, which has the correct behavior of returning a new object on each call. This is, unfortunately, a breaking change, so I do not want to make it without feedback. ", 
    "reporter": "cweaver", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1550067295757382", 
    "component": "combo reconstruction", 
    "summary": "common_variables.direct_hits.default_definitions is dangerous", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2017-03-18T16:56:17", 
    "milestone": "", 
    "owner": "mwolf", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20170318T17:10:25: david.schultz commented:

> I agree with cweaver about the global state.  Either users need to know that they need to make a deepcopy before appending (maybe add that directly to the error message), or the functional form should be used exclusively.  Personally, I'd go with the functional form for the API.

**Comment by icecube-trac** [comment]
Trac update at 20170422T06:04:54: kjmeagher changed milestone from "" to "IceRec 5.2"

**Comment by icecube-trac** [comment]
Trac update at 20170911T17:17:05: 
* kjmeagher changed status from "new" to "closed"
* kjmeagher commented:

> In changeset 158052:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="158052"
> replace all refrences to `direct_hits.default_definitions` to `direct_hits.get_default_definitions()` to avoid issues with global variables. Also add better checks to I3DirectHitsCalculator. fixes #1959
> ```


* kjmeagher changed resolution from "" to "fixed"

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:14:55: 
* olivas changed milestone from "IceRec 5.2" to ""
* olivas commented:

> Milestone IceRec 5.2 deleted
